src/NSSolverStationary.cpp

```diff
@@ -0,0 +1,933 @@
+    solution_owned.reinit(block_owned_dofs, MPI_COMM_WORLD);
+    delta_owned.reinit(block_owned_dofs, MPI_COMM_WORLD);
+
+    solution.reinit(block_owned_dofs, block_relevant_dofs, MPI_COMM_WORLD);
+  }
+}
+
+void NSSolverStationary::assemble_system(bool global_first_iter, bool computing_stokes)
+{
+  const unsigned int dofs_per_cell = fe->dofs_per_cell;
+  const unsigned int n_q = quadrature->size();
+  const unsigned int n_q_face = quadrature_face->size();
+
+  FEValues<dim> fe_values(*fe,
+                          *quadrature,
+                          update_values | update_gradients |
+                              update_quadrature_points | update_JxW_values);
+  FEFaceValues<dim> fe_face_values(*fe,
+                                   *quadrature_face,
+                                   update_values | update_normal_vectors |
+                                       update_JxW_values);
+
+  FullMatrix<double> cell_matrix(dofs_per_cell, dofs_per_cell);
+  FullMatrix<double> cell_pressure_mass_matrix(dofs_per_cell, dofs_per_cell);
+  Vector<double> cell_rhs(dofs_per_cell);
+
+  std::vector<types::global_dof_index> dof_indices(dofs_per_cell);
+
+  jacobian_matrix = 0.0;
+  residual_vector = 0.0;
+  pressure_mass = 0.0;
+
+  // Extractors
+  FEValuesExtractors::Vector velocity(0);
+  FEValuesExtractors::Scalar pressure(dim);
+
+  // We use the following vectors to store the old solution (i.e. at previous Newton
+  // iteration) and its gradient on quadrature nodes of the current cell.
+  std::vector<Tensor<1, dim>> velocity_loc(n_q);
+  std::vector<Tensor<2, dim>> velocity_gradient_loc(n_q);
+  std::vector<double> pressure_loc(n_q);
+
+  // tensor used to store the nonlinear term contribution in each quadrature point
+  // useful when computing (u . nabla) uv, corresponding to c(u;u,v) term
+  Tensor<1, dim> nonlinear_term;
+
+  for (const auto &cell : dof_handler.active_cell_iterators())
+  {
+    if (!cell->is_locally_owned())
+      continue;
+
+    fe_values.reinit(cell);
+
+    cell_matrix = 0.0;
+    cell_rhs = 0.0;
+    cell_pressure_mass_matrix = 0.0;
+
+    // We need to compute the Jacobian matrix and the residual for current
+    // cell. This requires knowing the value and the gradient of u^{(k)}
+    // (stored inside solution) on the quadrature nodes of the current
+    // cell. This can be accomplished through
+    // FEValues::get_function_values and FEValues::get_function_gradients.
+    fe_values[velocity].get_function_values(solution, velocity_loc);
+    fe_values[velocity].get_function_gradients(solution,
+                                               velocity_gradient_loc);
+    fe_values[pressure].get_function_values(solution, pressure_loc);
+
+    for (unsigned int q = 0; q < n_q; ++q)
+    {
+      for (unsigned int i = 0; i < dofs_per_cell; ++i)
+      {
+        for (unsigned int j = 0; j < dofs_per_cell; ++j)
+        {
+          if (global_first_iter || computing_stokes)
+          {
```

Comment:
Due to the similarity between Stokes and Navier-Stokes, this and the other branch have quite a few common terms (basically, all terms except for the transport term). It might be beneficial for the organization and clarity of the code to keep those common parts together.